packages/tizen_rpc_port/example/client/lib/message_client.dart

```diff
@@ -0,0 +1,211 @@
+// Copyright 2022 Samsung Electronics Co., Ltd. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// ignore_for_file: public_member_api_docs
+
+import 'package:tizen_rpc_port/tizen_rpc_port.dart';
+
+const String _tidlVersion = '1.9.1';
+
+enum _DelegateId {
+  notifyCallback(1);
+
+  const _DelegateId(this.id);
+  final int id;
+}
+
+enum _MethodId {
+  result(0),
+  callback(1),
+  register(2),
+  unregister(3),
+  send(4);
+
+  const _MethodId(this.id);
+  final int id;
+}
+
+abstract class _CallbackBase extends Parcelable {
+  _CallbackBase(this.id, this.once) {
+    sequenceId = sequenceNum++;
+  }
+
+  int id = 0;
+  bool once = false;
+  int sequenceId = 0;
+  static int sequenceNum = 0;
+
+  String get tag => '$id::$sequenceId';
+
+  Future<void> _onReceivedEvent(Parcel parcel);
+
+  @override
+  void serialize(Parcel parcel) {
+    parcel.writeInt32(id);
+    parcel.writeInt32(sequenceId);
+    parcel.writeBool(once);
+  }
+
+  @override
+  void deserialize(Parcel parcel) {
+    id = parcel.readInt32();
+    sequenceId = parcel.readInt32();
+    once = parcel.readBool();
+  }
+}
+
+abstract class NotifyCallback extends _CallbackBase {
+  NotifyCallback({bool once = false})
+      : super(_DelegateId.notifyCallback.id, once);
+
+  Future<void> onReceived(String sender, String message);
+
+  @override
+  Future<void> _onReceivedEvent(Parcel parcel) async {
+    final String sender = parcel.readString();
+    final String message = parcel.readString();
+    await onReceived(sender, message);
+  }
+}
+
+abstract class Message extends ProxyBase {
```

Comment:
Would it be necessary to make `Message` an abstract class? c.f. In gRPC, you can create an instance of the proxy class directly from the generated code and invoke any method without extending/overriding any class/method.

Reply:
It's the intentional design to receive the connected/disconnected/rejected events.

Reply:
I think it's okay to implement it as a normal class and make some callbacks empty functions.
If you'd like, I'll change it that way.
Thank you.

Reply:
@upple Yes, that's what I meant. It doesn't have to an abstract class. You may also update the example app to use `Message` as it is, without defining the `MyMessageClient` class. That will make the state management easier.

Reply:
For example, if you want to show the "Connected" message after a connection is established, you can just write

```dart
Message _client;

@override
void initState() {
  super.initState();

  _client.connect().then((_) {
    setState(() {
      _message = 'Connected';
    });
  }).catchError((dynamic error) {
    setState(() {
      _message = 'Connection failed: $error';
    });
  });
}
```

Reply:
By the way, if `connect` is called from `initState`, the app has to be restarted to reconnect if the connection fails (e.g. when the client app launched before the server app starts listening for connection). Since the "implicit connect" has not been implemented, you may need another button widget for "Connect".

Reply:
No... you're not understanding me correctly.

> First of all, the inheritance of the event callback is required because the `connect()` is asynchronous and other methods can be used after receiving the event.

It's just a matter of your app design. I already presented you sample code in the above. If you want to make sure `register` is called after `connect`, you can for example write

```dart
await _client.connect();
await _client.register(...);
```

or even add some UI that indicates the connection status.

> In addition, the current implementation is difficult to implement without inheritance because the status is updated through `MessageReceiver`.

Did you read my sample code in the above? Also you don't need `MessageReceiver` at all if you follow my [above suggestion](https://github.com/flutter-tizen/plugins/pull/438#discussion_r987716693). The `_MyAppState` class can directly extend the `NotifyCallback` interface.

Reply:
Wait.. does `connect()` return before the connection handshake is completed? It doesn't quite make sense to me and looks like a design fault that needs a fix. You may have to use Dart's [Completer](https://api.flutter.dev/flutter/dart-async/Completer-class.html) which is similar to C#'s TaskCompletionSource (which helps converting Event-based Asynchronous Pattern code to Task-based Asynchronous Pattern).

Reply:
Thank you for your review.
I understand your opinion.
It seems to be a different configuration from the existing native and c# implementations.
If you modify it like that, you will actually not need `onConnectedEvent` and `onRejectedEvent` callbacks.
Do you want to modify it in this direction?
In this case, a lot of things are likely to change.
If you don't mind, could you reflect the fix in the next release? (7.5)

Reply:
Basically, the connect() method is sending the connecion request asynchronously.
To avoid deadlock issues, we implmented it with async API at first time.
(That's why the event methods exist.)

How about modifying the description of the connect() method?

Reply:
If the implementation of the connect() method is changed to using the tizen.rpc_port_proxy_connect_sync() instead of the tizen.rpc_port_proxy_connect(), we can remove the event callbacks.

Reply:
> If you modify it like that, you will actually not need `onConnectedEvent` and `onRejectedEvent` callbacks.
Do you want to modify it in this direction?

Yes. That's what I suggested. (You will have to consider how to handle the "disconnected" event.)

> In this case, a lot of things are likely to change.

Not really.

> If the implementation of the connect() method is changed to using the tizen.rpc_port_proxy_connect_sync() instead of the tizen.rpc_port_proxy_connect(), we can remove the event callbacks.

What.. I didn't say you would have to use synchronous APIs or remove event streams. Callbacks, events (Event or Stream), and async/await (Task or Future) are all just different ways of implementing asynchronism, where callbacks are typically used by low-level languages like C and async/await is only supported by modern languages like C# and Dart. In general, you shouldn't mix any two of them for implementing a single feature. Your current code has this problem.

Event-based asynchronous code can be converted into Task (async/await)-based asynchronous code by using TaskCompletionSource (or Dart's Completer) as I mentioned above. You might refer to the "Wrapping event-based APIs" section of [my this article](https://developer.samsung.com/tizen/blog/en-us/2020/06/30/asynchronous-coding-concepts-for-tizen-net-applications#4.-Wrapping-event-based-APIs) for details. If a connection request is "rejected" by the server, you might throw a custom exception like `ConnectionRejectedException` using `Completer.completeError`.

Alternatively, callback-based asynchronous code (your current C++ implementation) can be wrapped into async/await-based Dart API by simply sending a method channel result after a "connected" or "rejected" callback is invoked. An example of such code can be found in other plugins like [geolocator_tizen](https://github.com/flutter-tizen/plugins/blob/70047edf2d33485469292c6791609c3a229c54a0/packages/geolocator/tizen/src/geolocator_tizen_plugin.cc#L230-L243).

I think both ways are feasible. If you're familiar with C++ and method channel, you may prefer the latter. Otherwise you can choose the former.

Reply:
> 

Thank you for the explanation.
I think we have to use Dart's Completer. It's better than current implementation.
If we implement connect() using Dart's Completer, we can hide it using _ of onConnectedEvent and onRejectedEvent. 
I think it would be more usable to implement it like this for app developers.

Reply:
Thanks. If you run into any problem or need help, please let me know.

Reply:
@swift-kim 
Thank you for your suggestion.
We refactored `ProxyBase` as you advised.
It seems to be a better design than before.

Please check the new patch. 
I hope you like it.

Reply:
The direction seems correct but the code needs more improvements. Especially you should reduce the number of unnecessarily overridden methods. Also don't forget that you're supposed to [migrate the NotifyCallback class into an interface](https://github.com/flutter-tizen/plugins/pull/438#discussion_r987716693).

Reply:
upple is going to change the implementation to the following code:
```
typedef NotifyCallback = void Function(String, String);

class Message {
  int register(String name, NotifyCallback callback) {
    ...
  }
}
```

I've tested it, it works fine and the application code is simple.